Container based deployment

aravindavk opened the pull request:
* Two containers are built. `kadalu-storage/server` and
  `kadalu-storage/node`
* Start one instance of `kadalu-storage/server` and multiple
  instances of `kadalu-storage/node` as required.
* Refer container-README.md for build and usage instructions

Signed-off-by: Aravinda Vishwanathapura <aravinda@kadalu.io>

aravindavk committed:
Container based deployment

* Two containers are built. `kadalu-storage/server` and
  `kadalu-storage/node`
* Start one instance of `kadalu-storage/server` and multiple
  instances of `kadalu-storage/node` as required.
* Refer container-README.md for build and usage instructions

Signed-off-by: Aravinda Vishwanathapura <aravinda@kadalu.io>

amarts commented:
Nice improvements

amarts commented on container-README.md:
```diff
@@ -0,0 +1,87 @@
+* Add Volumes for `/var/lib/kadalu` for each Node containers
+* Add Volume for `kadalu-storage/server`(For Sqlite db and other things)
+* Set the required environment variables
+* How to communicate between containers(Self-heal daemon requires to connect to other storage nodes, kadalu-node agent needs to connect to Kadalu Server/Control plane)
```
what happens if this is communicated using 'external' IP even though for 1 brick it can talk locally? As the config object to generate the volfile would be containing the information about the node level (read as ports etc).

aravindavk replied:
Found this method to create a network that works across the nodes. I am yet to try  https://www.youtube.com/watch?v=nGSNULpHHZc

For local containers, all are now part of same network

amarts commented on container-README.md:
```diff
@@ -0,0 +1,87 @@
+* Access brick directories inside `kadalu-storage/node` container(Dynamic access when new brick added)
+* Add volume for log files or use syslog?
+* Set Proper Hostname for `kadalu-storage/node` containers.
+* Changes required in Dockerfile to build for Arm (Refer [build_arm64_static.sh](./build_arm64_static.sh))
```
can keep for later to start with.

amarts commented on Dockerfile.node:
```diff
@@ -0,0 +1,58 @@
+COPY --from=crystal /app/node/bin/kadalu-node /usr/sbin/kadalu-node
+COPY --from=crystal /app/cli/bin/kadalu /usr/sbin/kadalu
+
+RUN systemctl enable kadalu-node
```
asking bit late I guess. Which is quicker? kadalu-node as binary handling glusterfs docker images? or a single docker image with node and everything combined?

aravindavk replied:
Found this approach simple. With this approach we can spawn multiple containers in the same machine as required. each container will become node/brick. It is also possible to export multiple bricks from same container.

amarts commented on container-README.md:
```diff
@@ -0,0 +1,87 @@
+* Add Volume for `kadalu-storage/server`(For Sqlite db and other things)
+* Set the required environment variables
+* How to communicate between containers(Self-heal daemon requires to connect to other storage nodes, kadalu-node agent needs to connect to Kadalu Server/Control plane)
+* Access brick directories inside `kadalu-storage/node` container(Dynamic access when new brick added)
```
can we start with assumption (or even a restriction) saying 1 node docker image has just one brick? 

Considering this would allow multiple kadalu-nodes to run in single machine, thats not a bad assumption to start with, and would allow us to not look at dynamic part of this.

aravindavk replied:
ack. It should be possible to add multiple volumes while starting node containers.

amarts commented on container-README.md:
```diff
@@ -0,0 +1,87 @@
+* Set the required environment variables
+* How to communicate between containers(Self-heal daemon requires to connect to other storage nodes, kadalu-node agent needs to connect to Kadalu Server/Control plane)
+* Access brick directories inside `kadalu-storage/node` container(Dynamic access when new brick added)
+* Add volume for log files or use syslog?
```
do we need a volume? or if volume is present (with `-v` option) then use it, otherwise, let people know that logs would be lost?

aravindavk replied:
ack

aravindavk committed:
server: Fix database path

Signed-off-by: Aravinda Vishwanathapura <aravinda@kadalu.io>

aravindavk committed:
Fix issues after upgrading to Crystal 0.36.1

Signed-off-by: Aravinda Vishwanathapura <aravinda@kadalu.io>

aravindavk committed:
Add details of container volumes

And fix an issue of missing directories when `/var/lib/kadalu` is
added as Volume.

Signed-off-by: Aravinda Vishwanathapura <aravinda@kadalu.io>

aravindavk commented:
Will merge this PR now, basic things are now working. Found a separate issues related to volume start, will send fix as separate PR.
